### ISSUE TITLE
more convenient way to use a fixed seed for a test group

### ISSUE THREAD

**Issue by ghorn** [hedgehogqa/haskell-hedgehog#201]
I would like to use our extensive hedgehog test suite on my continuous integration server, but fortunately and unfortunately it occasionally finds new corner cases and fails commits which have nothing to do with the hedgehog tests so I had to disable them.

The solution seems to be to always run with the same seed on CI to prevent regressions but not find new issues, and then some mega nightly build with a random seed to look for new failures.

The problem is the only seeded test function I could find is `recheck` which serves a different purpose. Is there a reason that `Hedgehog.Internal.Runner.checkGroupWith` can't take a seed? I could jigger `recheck` to do this but I would be missing out on all the advantages of using `checkGroup`.

**Comment by ghorn** [comment]
If someone could provide a few pointers, I'd be happy to do this work.

**Comment by thumphries** [comment]
Hey! Sorry for the slow.

We should be able to support this workflow with a couple of changes. If you float a `Maybe Seed` up into the signature for `checkGroupWith`, that is probably a good start. We can bikeshed the API change once it's working.

Currently each property gets a new seed in [`checkNamed`](https://github.com/hedgehogqa/haskell-hedgehog/blob/a506b6d2ec9b8721aead61b72a2f1fa8b4c88ce5/hedgehog/src/Hedgehog/Internal/Runner.hs#L238), which calls `checkRegion`, which takes a `Seed`. I'm not too sure what the behaviour should be with a fixed seed, but you might want to split it before each property.

It's a little outside the way most people use Hedgehog, usually the continuous search for bugs in CI is a feature. But I see where you're coming from and we should have a better story. Anecdotally I have heard of teams discouraged from doing property-based testing in monorepo situations, which is pretty terrible!

**Comment by thumphries** [comment]
An alternative design: read `HEDGEHOG_SEED` from `ENV` and use that if defined. Unsure if wise

**Comment by ghorn** [comment]
Thanks for the pointers. I strongly prefer the API change over the environment variable. Let me give it a try.

**Comment by moodmosaic** [comment]
Here's a patch for setting a fixed seed through the `HEDGEHOG_SEED` environment variable, which has to be in the form of `15417483394291888538 15013407834460730181`.

```diff
diff --git a/hedgehog/src/Hedgehog/Internal/Config.hs b/hedgehog/src/Hedgehog/Internal/Config.hs
index c679959..efd3320 100644
--- a/hedgehog/src/Hedgehog/Internal/Config.hs
+++ b/hedgehog/src/Hedgehog/Internal/Config.hs
@@ -8,6 +8,9 @@ module Hedgehog.Internal.Config (
     UseColor(..)
   , resolveColor
 
+  , Seed(..)
+  , resolveSeed
+
   , Verbosity(..)
   , resolveVerbosity
 
@@ -16,14 +19,20 @@ module Hedgehog.Internal.Config (
 
   , detectMark
   , detectColor
+  , detectSeed
   , detectVerbosity
   , detectWorkers
   ) where
 
 import           Control.Monad.IO.Class (MonadIO(..))
 
+import qualified Data.Text as Text
+
 import qualified GHC.Conc as Conc
 
+import           Hedgehog.Internal.Seed (Seed(..))
+import qualified Hedgehog.Internal.Seed as Seed
+
 import           Language.Haskell.TH.Lift (deriveLift)
 
 import           System.Console.ANSI (hSupportsANSI)
@@ -106,6 +115,20 @@ detectColor =
           else
             pure DisableColor
 
+detectSeed :: MonadIO m => m Seed
+detectSeed =
+  liftIO $ do
+    menv <- lookupEnv "HEDGEHOG_SEED"
+    case menv of
+      Nothing ->
+        Seed.random
+      Just env ->
+        let
+          [value, gamma] =
+            read . Text.unpack <$> Text.splitOn (Text.pack " ") (Text.pack env)
+        in
+          pure $ Seed value gamma
+
 detectVerbosity :: MonadIO m => m Verbosity
 detectVerbosity =
   liftIO $ do
@@ -141,6 +164,13 @@ resolveColor = \case
   Just x ->
     pure x
 
+resolveSeed :: MonadIO m => Maybe Seed -> m Seed
+resolveSeed = \case
+  Nothing ->
+    detectSeed
+  Just x ->
+    pure x
+
 resolveVerbosity :: MonadIO m => Maybe Verbosity -> m Verbosity
 resolveVerbosity = \case
   Nothing ->
diff --git a/hedgehog/src/Hedgehog/Internal/Runner.hs b/hedgehog/src/Hedgehog/Internal/Runner.hs
index 6c71f42..fdb8fcc 100644
--- a/hedgehog/src/Hedgehog/Internal/Runner.hs
+++ b/hedgehog/src/Hedgehog/Internal/Runner.hs
@@ -236,10 +236,11 @@ checkNamed ::
   => Region
   -> Maybe UseColor
   -> Maybe PropertyName
+  -> Maybe Seed
   -> Property
   -> m (Report Result)
-checkNamed region mcolor name prop = do
-  seed <- liftIO Seed.random
+checkNamed region mcolor name mseed prop = do
+  seed <- resolveSeed mseed
   checkRegion region mcolor name 0 seed prop
 
 -- | Check a property.
@@ -247,7 +248,7 @@ checkNamed region mcolor name prop = do
 check :: MonadIO m => Property -> m Bool
 check prop =
   liftIO . displayRegion $ \region ->
-    (== OK) . reportStatus <$> checkNamed region Nothing Nothing prop
+    (== OK) . reportStatus <$> checkNamed region Nothing Nothing Nothing prop
 
 -- | Check a property using a specific size and seed.
 --
@@ -275,8 +276,9 @@ checkGroup config (Group group props) =
 #endif
     putStrLn $ "━━━ " ++ unGroupName group ++ " ━━━"
 
+    seed <- resolveSeed Nothing
     verbosity <- resolveVerbosity (runnerVerbosity config)
-    summary <- checkGroupWith n verbosity (runnerColor config) props
+    summary <- checkGroupWith n verbosity (runnerColor config) seed props
 
     pure $
       summaryFailed summary == 0 &&
@@ -291,9 +293,10 @@ checkGroupWith ::
      WorkerCount
   -> Verbosity
   -> Maybe UseColor
+  -> Seed
   -> [(PropertyName, Property)]
   -> IO Summary
-checkGroupWith n verbosity mcolor props =
+checkGroupWith n verbosity mcolor seed props =
   displayRegion $ \sregion -> do
     svar <- atomically . TVar.newTVar $ mempty { summaryWaiting = PropertyCount (length props) }
 
@@ -331,7 +334,7 @@ checkGroupWith n verbosity mcolor props =
     summary <-
       fmap (mconcat . fmap (fromResult . reportStatus)) $
         runTasks n props start finish finalize $ \(name, prop, region) -> do
-          result <- checkNamed region mcolor (Just name) prop
+          result <- checkNamed region mcolor (Just name) (Just seed) prop
           updateSummary sregion svar mcolor
             (<> fromResult (reportStatus result))
           pure result
```

See if it works for you, and if it does, let's discuss about taking that in instead of #234. And perhaps there are other, easier ways of doing this. /cc @jystic @thumphries

**Comment by moodmosaic** [comment]
@peddie, @thumphries, here's some sample usage:

```haskell
import           Control.Monad.IO.Class (MonadIO(..))

import           Hedgehog
import qualified Hedgehog.Gen as Gen
import qualified Hedgehog.Range as Range

import           Hedgehog.Internal.Runner

checkSequentialWith :: MonadIO m => RunnerConfig -> Group -> m Bool
checkSequentialWith config =
  checkGroup config

tests :: IO Bool
tests =
  checkSequentialWith
    RunnerConfig {
        runnerWorkers =
          Nothing
      , runnerColor =
          Nothing
      , runnerVerbosity =
          Nothing
      , runnerSeed =
          Just $ Seed 123 456
      }
    $$(discover)
```

That, or just set the `HEDGEHOG_SEED` environment variable.

---

This pretty much follows the same pattern used for the rest of the stuff in `RunnerConfig`. See [updated diff](https://github.com/hedgehogqa/haskell-hedgehog/issues/201#issuecomment-441401915) above.

**Comment by peddie** [comment]
Even though it's `Internal`, with the ability to set the seed directly via the `RunnerConfig`, I'd be happy with this diff.

**Comment by ghorn** [comment]
That works for me.

**Comment by simfleischman** [comment]
Are there any objections to applying this patch in a PR to the master branch?

**Comment by moodmosaic** [comment]
Not that I know of 👍

**Comment by simfleischman** [comment]
Great, #424 is ready to go from my perspective. I'm happy to change as needed.

**Comment by jacobstanley** [comment]
Merged #446 (which is an update of #424)

**Comment by jacobstanley** [comment]
Available in [hedgehog-1.1.1](https://hackage.haskell.org/package/hedgehog-1.1.1)
